lizard_auth_server/views.py

```diff
@@ -304,3 +324,143 @@ def profile(self):
+            return True
+        except ValidationError:
+            return False
+
+    @staticmethod
+    def is_portal(sso_key):
+        """Validate a portal.
+
+        Args:
+            sso_key (str): a public key identifying the portal.
+
+        Returns:
+            bool: True if the portal exists, False otherwise.
+
+        """
+        return Portal.objects.filter(sso_key=sso_key).exists()
+
+    @staticmethod
+    def get_token(user, portal, exp=None):
+        """Return a JSON Web Token.
+
+        Args:
+            user (django.contrib.auth.models.User): an active, authenticated
+                user that has access to the portal.
+            portal (lizard_auth_server.models.Portal): a Portal instance.
+            exp (int/datetime): expiration time - in UTC - on or after
+                the token must not be accepted for processing.
+
+        Returns:
+            str: A JSON Web Token.
+
+        """
+        assert(user.is_active and user.is_authenticated())
+        assert(user.user_profile.has_access(portal))
+        if exp is None:
+            exp = datetime.utcnow() + JWT_EXPIRATION_DELTA
+        payload = {'exp': exp, 'username': user.username}
+        secret = portal.sso_secret
+        token = jwt.encode(payload, secret, algorithm=JWT_ALGORITHM)
+        return token
+
+    @method_decorator(login_required)
+    def dispatch(self, *args, **kwargs):
+        return super(JWTView, self).dispatch(*args, **kwargs)
+
+    def get(self, request, *args, **kwargs):
+        """Return a token.
+
+        If the request has a `next` query string parameter, the response will
+        be a redirect (302) and the token is added to the `next` URL. If a
+        `next` parameter is absent, a JSON response is returned.
+
+        An HTTP 400 Bad Request is returned in case of trouble.
+
+        """
+        self.redirect_to = request.GET.get('next', '')
+        sso_key = request.GET.get('portal', '')
+
+        if self.redirect_to and not JWTView.is_url(self.redirect_to):
+            reason = _('Invalid `next` query string parameter.')
+            return HttpResponseBadRequest(reason, content_type='text/plain')
+
+        if not sso_key:
+            reason = _('Missing `portal` query string parameter.')
+            return HttpResponseBadRequest(reason, content_type='text/plain')
+
+        if not JWTView.is_portal(sso_key):
+            reason = _('Invalid `portal` query string parameter.')
+            return HttpResponseBadRequest(reason, content_type='text/plain')
+
+        portal = Portal.objects.get(sso_key=sso_key)
+
+        if not request.user.user_profile.has_access(portal):
+            reason = _('You do not have access to this portal.')
+            return HttpResponseBadRequest(reason, content_type='text/plain')
+
+        self.token = JWTView.get_token(request.user, portal)
+
+        if self.redirect_to:
+            return redirect(self.success_url)
+        else:
+            return self.as_json()
```

Comment:
Dit vraag ik me nog af: waarom een json teruggeven in plaats van alleen het token zelf, als string, teruggeven? Misschien wel handig, hoor, om het in json te verpakken.